src/java.base/share/classes/java/util/ArrayList.java

```diff
@@ -750,6 +750,22 @@ public void clear() {
      * @throws NullPointerException if the specified collection is null
      */
     public boolean addAll(Collection<? extends E> c) {
+        if (c.getClass() == ArrayList.class) {
+            ArrayList<?> src = (ArrayList<?>) c;
+            Object[] a = src.elementData;
+            int numNew = src.size;
+            if (numNew == 0)
+                return false;
```

Comment:
Previously `addAll()` with an empty list would increment `modCount`, whereas now it won't. Might want to move this test down after `modCount++` to avoid the unnecessary behavioral change.